### PR TITLE
Support Granite 3.0 and 3.1 models

### PR DESCRIPTION
**Pull request by JamesKunstle**
Granite 3.(0,1) models are Llama-architecture models with some different scaling terms in various places. This commit adds granite model patching for decoder-only granite 3 models (not multimodal) and the corresponding tests.

## Summary
This change enables patching Granite 3.(0,1) models w/ Liger kernels. We would like to use Liger kernels [in our training implementation](https://github.com/instructlab/training) but we're a Granite-first codebase for the moment.

## Testing Done
Convergence tests confirm that loss and model parameters are equivalent w/ and w/o Liger kernels. Logits, however, are not equivalent even when only swapping the SwiGLUMLP layer. The ator and rtol may need to be tuned for Granite vs. Llama, I'm going to continue investigating this before this PR is merged.

- Hardware Type: EC2 g6e.12xlarge; 4xL40s
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by JamesKunstle** [comment]
Fixes #557

**Review by DRXD1000 — COMMENTED** [review]
Conversion Tests (except for Multimodal ones) have been done with the changes in [here](https://github.com/JamesKunstle/Liger-Kernel/pull/1)

**Comment by JamesKunstle** [comment]
@DRXD1000 Thank you so much for the support! I really appreciate that you caught that config PEBCAK, I've been trying to figure out why the layer has different output behavior from Llama!

**Review by DRXD1000 — APPROVED** [review]
I would approve, despite a test error in dpo_loss since this has not been touched here.

**Comment by DRXD1000** [comment]
@JamesKunstle was a pleasure! I can't see what causes the merge conflict, if i can be of further assistance let  me know :)

**Comment by JamesKunstle** [comment]
@DRXD1000 For my future education: how did you debug the logit scaling value problem and pick the values per data type?

**Comment by DRXD1000** [comment]
@JamesKunstle i was looking at the granite implementation in huggingface and tried to load the model with Llama directly. After a short benchmark i noticed this did not work. 

After reading the modeling_granite.py in transformers again i noticed the logits_scaling with the very obvious #Main difference  to llama comment (must have been blind the first time reading it...) 

After that i checked your pr, changed the Tests and ran them. The test failed at logits. With the source Code in mind I then looked at the default settings in GraniteConfig and at some values of the trained models. Since they scaled with model size i gradually increased them till the test passed. 

Since normal users will not do pretraing and a proper value of logits_scaling will allready exists, it was fine for me to do this trial and error approach. 

Wow this got longer than planned 😅

**Comment by JamesKunstle** [comment]
@DRXD1000 That's an excellent explanation, thanks a lot, it helps a lot! I hadn't considered that value- I figured it was defaulted in the config so I didn't inspect it. I was trying to debug by isolating the layer from the `GraniteMLP` layer and comparing the individual logits and _those_ were different too in my testing, so I was pretty confused. This seems like a much better way to investigate.

**Comment by JamesKunstle** [comment]
Sorry for any reviewers- `ruff` reformatted lots of stuff from `make checkstyle`.

**Comment by JamesKunstle** [comment]
@ByronHsu Would like to have the workflow tests approved for merge!

**Review by lancerts — APPROVED** [review]

**Comment by hebiao064** [comment]
@JamesKunstle Thanks for the contribution.
However, our `make test-convergence` test is failing now https://github.com/linkedin/Liger-Kernel/actions/runs/13464011417/job/37625671758?pr=578
Since Granite wasn't introduced in transformers 4.44. 

Would you mind share the context why do you want to add this into convergence test?

cc @lancerts @shivam15s

**Comment by hebiao064** [comment]
Screenshot 2025-02-21 at 12 10 53 PM
https://github.com/huggingface/transformers/commit/c35d2ccf5a5ebcab24a0b9da3cb2527b3ae565a0
It's added by transformer 4.45

**Comment by shivam15s** [comment]
Blocked convergence tests for Granite for transformers backward compatibility tests in #578
